src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1192,401 @@ where
+        let ltree_path = LTree::from(traversal_path.iter());
+        // The created field in the table is for the leaf height i.e block number
+        let created = block_number as i64;
+        let mut txn = self.transaction().await?;
+
+        // Insert the leaf hash into a separate table and use the id referenced by the state table
+        // This avoids redundancy of hashes in the state table to save space
+        let insert_hash = txn
+            .query_one(
+                "INSERT INTO hash(value) VALUES ($1) 
+                ON CONFLICT (value) DO UPDATE SET value = excluded.value 
+                RETURNING id",
+                &[&leaf_commit],
+            )
+            .await?;
+        let hash_id: i32 = insert_hash.get(0);
+
+        // upsert the leaf node
+        txn.upsert(
+            name,
+            ["pos", "created", "hash_id", "index", "entry"],
+            ["pos", "created"],
+            [[
+                sql_param(&ltree_path),
+                sql_param(&created),
+                sql_param(&hash_id),
+                sql_param(&serde_json::to_value(&pos).map_err(ParseError::Serde)?),
+                sql_param(&serde_json::to_value(elem).map_err(ParseError::Serde)?),
+            ]],
+        )
+        .await?;
+
+        let mut traversal_path = traversal_path.into_iter();
+
+        // Insert the internal nodes using the membership proof.
+        for node in proof.merkle_path().iter() {
+            match node {
+                MerkleNode::Branch { value, children } => {
+                    // Get hash
+                    let mut writer = [0_u8; 32];
+                    value
+                        .serialize_compressed(&mut writer[..])
+                        .map_err(ParseError::Serialize)?;
+
+                    // Insert the internal node hash value
+                    let row = txn
+                        .query_one(
+                            "INSERT INTO hash(value) VALUES ($1) 
+                            ON CONFLICT (value) DO UPDATE SET value = excluded.value
+                            RETURNING id",
+                            &[&writer.to_vec()],
+                        )
+                        .await?;
+
+                    let hash_id: i32 = row.get(0);
+                    let mut children_values = Vec::new();
+                    // We only insert the non-empty children in the children field of the table
+                    // BitVec is used to separate out Empty children positions
+                    let mut children_bitvec = BitVec::new();
+                    for child in children {
+                        let child = child.as_ref();
+                        match child {
+                            MerkleNode::Empty => {
+                                children_bitvec.push(false);
+                            }
+                            MerkleNode::Branch { .. } => (),
+                            MerkleNode::Leaf { .. } => (),
```

Comment:
It seems like we would need to push a `true` bit in these cases, but I think the reason why it's ok is because these cases are never reached, since in a merkle path the children are always forgotten. Still, I think we should either handle these cases properly or return an error, not silently ignore them